src/theory/splitter.h

```diff
@@ -0,0 +1,136 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Amalee Wilson, Andrew Wu
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Splitter for creating partitions.
+ */
+
+#include "cvc5_private.h"
+
+#ifndef CVC5__THEORY__SPLITTER_H
+#define CVC5__THEORY__SPLITTER_H
+
+#include <fstream>
+#include <list>
+#include <sstream>
+
+#include "proof/trust_node.h"
+#include "theory/valuation.h"
+
+namespace cvc5 {
+
+class TheoryEngine;
+
+namespace prop {
+class PropEngine;
+}
+
+namespace theory {
+
+class Splitter
+{
+ public:
+  Splitter(TheoryEngine* theoryEngine, prop::PropEngine* propEngine);
+
+  /**
+   * Make partitions for parallel solving.
+   */
+  TrustNode makePartitions();
+
+ private:
+  /**
+   * Print the cube to the specified output.
+   */
+  void emitCube(Node toEmit);
+
+  /**
+   * Block a path in the search by sending the not of toBlock as a lemma to the
+   * SAT solver.
+   */
+  TrustNode blockPath(Node toBlock);
+
+  /**
+   * Close the output file if it is specified.
+   */
+  void closeFile();
+
+  /**
+   * Stop partitioning and return unsat.
+   */
+  TrustNode stopPartitioning();
+
+  /**
+   * Get the list of decisions from the SAT solver
+   */
+  void collectDecisionLiterals(std::vector<TNode>& literals);
+
+  /**
+   * Current propEngine.
+   */
+  prop::PropEngine* d_propEngine;
+
+  /**
+   * Valuation of the theory engine.
+   */
+  std::unique_ptr<Valuation> d_valuation;
+
+  /**
+   * The number of partitions requested through the compute-partitions option.
+   */
+  const unsigned d_numPartitions;
+
+  /**
+   * How long to wait in terms of the number of checks until creating the first
+   * partition.
+   */
+  unsigned d_numChecks;
+
+  /**
+   * The number of partitions that have been created.
+   */
+  unsigned d_numPartitionsSoFar;
+
+  /**
+   * The name of the output file to write the partitions.
+   */
+  std::string d_partitionFile;
+
+  /**
+   * The filestream for writing the partitions.
+   */
+  std::ofstream d_partitionFileStream;
+
+  /**
+   * The output stream: either std::cout or the filestream if an output file is
+   * specified.
+   */
+  std::ostream* d_output;
```

Comment:
I wonder whether we should rather use our `--output` infrastructure for this, instead of manually maintaining this manually. @ajreynol any thoughts?

Also, I think the current interface is not really safe: it seems I can create the splitter with a filename (creating the file stream and storing a pointer in `d_output`) and call `closeFile()`. Then I end up with a pointer to a closed file stream.

Reply:
Would it be okay to just make closeFile private? It's just a helper method to make the partitioning strategies less cluttered.

Reply:
Concerning the use of the `--output` infrastructure, is this a question of whether to allocate an output tag for printing partitions?

If so, there are several other features right now that have the same question. In particular, `--sygus-rr-synth` enables a module that then uses standard out. 

I've thought a lot about changing this so that e.g. there is an output tag for printing synthesized rewrite rules.

However, I feel it actually is best to use an option in these cases and then just pick up the main output stream as needed.
In other words, `-o X` should not impact the behavior of the solver, only what it outputs.

Reply:
Following this argument, printing the partitions **somewhere** would be part of the core solver behavior that should not change, and thus we should not use `--output` here... is this what you mean?

Reply:
So I'm saying that `--compute-partitions` is the option one should use to enable the printing.  This changes the behavior of the solver quite a bit obviously, since it is focused on computing the partition, not solving.

It would not make sense to say e.g. `-o partitions` instead.

Reply:
Okay, I see.